File: app/src/main/java/com/github/abhijitpparate/keeps/screen/login/presenter/login/LoginContract.java
package com.github.abhijitpparate.keeps.screen.login.presenter.login;

import android.support.annotation.StringRes;

import com.facebook.login.LoginResult;
import com.github.abhijitpparate.keeps.data.auth.Credentials;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.twitter.sdk.android.core.TwitterSession;

public interface LoginContract {

    interface View {

        String getEmail();
        String getPassword();

        void showHomeScreen();
        void showRegistrationScreen();
        void showGoogleSignInScreen();
        void showFacebookSignInScreen();
        void showTwitterSignInScreen();
        void showProgressIndicator(boolean show);

        void setPresenter(LoginContract.Presenter presenter);
        void makeToast(@StringRes int strId);
        void makeToast(String message);
    }

    interface Presenter {
        void attemptLogin(Credentials credentials);

        void onLoginClick();
        void onGoogleLoginClick();
        void onFacebookLoginClick();
        void onTwitterLoginClick();

        void onRegisterClick();

        void subscribe();
        void unsubscribe();

        void registerGoogleAccount(GoogleSignInAccount account);
        void registerFacebookAccount(LoginResult loginResult);
        void registerTwitterAccount(TwitterSession twitterSession);

    }
}

File: app/src/main/java/com/github/abhijitpparate/keeps/screen/login/presenter/login/LoginContract.kt
package com.github.abhijitpparate.keeps.screen.login.presenter.login

import android.support.annotation.StringRes

import com.facebook.login.LoginResult
import com.github.abhijitpparate.keeps.data.auth.Credentials
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.twitter.sdk.android.core.TwitterSession

interface LoginContract {

    interface View {

        val email: String
        val password: String

        fun showHomeScreen()
        fun showRegistrationScreen()
        fun showGoogleSignInScreen()
        fun showFacebookSignInScreen()
        fun showTwitterSignInScreen()
        fun showProgressIndicator(show: Boolean)

        fun setPresenter(presenter: LoginContract.Presenter)
        fun makeToast(@StringRes strId: Int)
        fun makeToast(message: String)
    }

    interface Presenter {
        fun attemptLogin(credentials: Credentials)

        fun onLoginClick()
        fun onGoogleLoginClick()
        fun onFacebookLoginClick()
        fun onTwitterLoginClick()

        fun onRegisterClick()

        fun subscribe()
        fun unsubscribe()

        fun registerGoogleAccount(account: GoogleSignInAccount)
        fun registerFacebookAccount(loginResult: LoginResult)
        fun registerTwitterAccount(twitterSession: TwitterSession)

    }
}
